### DISCUSSION TITLE
Issue With Alarms

### DISCUSSION THREAD

**Discussion by bfailing** [artisan-roaster-scope/artisan#922] [Q&A] [answered]
The attached image shows a simple alarm setup to preheat the roaster for 5 minutes with no beans.  The problem is that it seems to ignore (or sees as met right away) the 5:00 from CHARGE alarm (number 7) and continues right along.  I understand that the "From" conditions are ignored with times of "0:00", but the 7th alarm has a time of 5:00.  And all the subsequent conditions are conditioned on the 7th alarm.  Why is this happening?

![Alarm Issue](https://user-images.githubusercontent.com/17439289/181470476-148d497a-d886-4ae7-b28d-5125e02dca1a.png)

**Comment by MAKOMO** [comment, marked as answer]
Each rule comes with a time and a temperature condition. Those conditions are disjunctive. If one of both holds, the rule "fires". Most likely your BT is above 0C at CHARGE and thus your rule 7 fires immediately. Just select the empty entry as temperature source for this rule and it will only "fire" once the time criteria is fulfilled.

**Reply by bfailing** [reply]
Thanks. So it sounds like the conditional statement for the alarms is:

IF: (If Alarm) AND NOT (But Not) AND [(From - Time) OR (Source - Condition - Value)]
THEN: (Action - Beep - Description)

Rather than OR-ing the (From - Time) and (Source - Condition - Value) terms, is it possible to AND them?

**Reply by poundy** [reply]
AND by creating two rules like you already do and set the IF ALARM...

**Reply by bfailing** [reply]
So I tried AND-ing alarms 10 and 11 in the attached image (below).  If alarm 10 triggers before alarm 11, everything seems to work fine.  However, if alarm 11 triggers before alarm number 10, it's as if alarm 10 is skipped and the fan is turned off before the "IT" temp gets down to 90 degrees.  Why is this happening?

My understanding is that alarm 11 is conditional on alarm 10.  So, if the conditions for alarm 11 are met before those of alarm 10, it will wait until alarm 10 is triggered to trigger alarm 11.  Is this not the case?

![Alarm Issue 2](https://user-images.githubusercontent.com/17439289/182811103-f36a60df-480d-47da-bc06-39b7d7a13a49.png)

**Reply by MAKOMO** [reply]
Alarm 11 cannot trigger before alarm 10 with this setup.

> My understanding is that alarm 11 is conditional on alarm 10. So, if the conditions for alarm 11 are met before those of alarm 10, it will wait until alarm 10 is triggered to trigger alarm 11. Is this not the case?

Alarm 11 is only evaluated after alarm 10 triggered.

Your alarm rule 10 has "From" set to "If Alarm" which results in its time/temp conditions to be interpreted relative to the point where the "If Alarm" rule it depends on is triggered, which is alarm rule 9. Your alarm rule 9 is trigged a second after DROP and set the fan to 80%. Thus your alarm rule 10 will trigger as soon as the temperature IT drops by 90 degree (after DROP), relative to the IT temperature at DROP.

**Reply by bfailing** [reply]
I tried removing the dependency of Alarm 10 on Alarm 9 (removed "9" from "If Alarm" dialog for Alarm 10) per your suggestion, but this resulted in the fan not being shut off via Alarm 11.  My understanding is that this would cause Alarm 10 to be triggered when the "IT" temp gets below 90 degrees (absolute temp, not relative).   

And I tried it both ways where Alarm 10 is met before Alarm 11 (as shown since IT drops to 90 degrees in about 1:00 after DROP which is below the 1:30 from DROP specified in Alarm 11) and vice versa (by changing the "Time" value of Alarm 11 to 0:30).  Same result in both cases.

Why is this happening?

**Reply by bfailing** [reply]
I figured it out.  I don't know why it was happening, but selecting "If Alarm" in the "From" dialog with a "0" in the "If Alarm" dialog seems to cause the alarm not to trigger.  Perhaps it is interpreted as relative to an alarm (number "0") that does not exist.  

So, I selected DROP in the "From" dialog for Alarm 10 and made Alarm 10 dependent on Alarm 9 as before.  It seems to work fine regardless of the order in which Alarms 10 and 11 are triggered.  

I'm not sure if selecting something other than DROP (and "If Alarm" of course) would also work.  I suspect so, but I didn't test it.

Below is the final set of alarms should it be helpful to anyone.

![Final Alarms For Preheat](https://user-images.githubusercontent.com/17439289/183224273-bf8a75da-6321-4d89-ba36-528e577a8fb4.png)
